In [25]:
import numpy as np #used for numerical calculation
import pandas as pd #used for handling dataset
from sklearn.model_selection import train_test_split#Splits data into training and testing sets. 
from sklearn.linear_model import LogisticRegression#Imports the Logistic Regression model for classification.
from sklearn.metrics import accuracy_score#Used to calculate the accuracy of the model’s predictions.

In [27]:
# loading the dataset to a Pandas DataFrame
df = pd.read_csv('creditcard.csv')

In [28]:
# first 5 rows of the dataset
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

Time:-represents time between first transaction and current transaction in sec
v1-v28 - principal components obtained using PCA,actual meaning is unkown due to privacy concerns
amount - represents the transaction amount

In [32]:
df.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [34]:
# dataset informations
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [36]:
# checking the number of missing values in each column
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [38]:
# distribution of legit transactions & fraudulent transactions
df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

This Dataset is highly unblanced
0 --> Normal Transaction

1 --> fraudulent transaction

The dataset is highly imbalanced because fraudulent transactions are very rare compared to legitimate ones


 If the model sees too many normal transactions, it might always predict "normal" and miss fraud.


In [81]:
# separating the data for analysis
legit = df[df.Class == 0]
fraud = df[df.Class == 1]

In [83]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


### statistical measures of the data
legit.Amount.describe()

1️ Understand Spending Patterns – Identifies the typical range of transaction amounts.
2️ Compare with Fraud Transactions – Detects differences in fraud vs. normal transactions.
3️ Detect Outliers – Finds unusually large or small amounts, which may indicate fraud.

In [88]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [90]:
# compare the values for both transactions
#Shows how fraud transactions differ from normal.
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Under-Sampling

Build a sample dataset containing similar distribution of normal transactions and Fraudulent Transactions

Number of Fraudulent Transactions -->
 492

randomly selects 492 legitimate transactions from the dataset.

Since there are 492 fraud cases, we take 492 legitimate transactions to balance the dataset.

In [95]:
legit_sample = legit.sample(n=492)

Concatenating two DataFrames

In [98]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)#if we give axis=0 added one after another(row)

In [100]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
153763  100005.0 -0.947856  0.488175  0.776310  3.244927  2.302845 -0.073014   
67879    52734.0  1.030633 -0.887742  1.446677  0.582385 -1.736129  0.030491   
190575  128890.0  1.722578 -0.055197  0.100595  3.725552 -0.225874  0.941944   
182469  125411.0 -0.196458  0.239535  0.155793 -2.081468  2.489614  3.927646   
216102  140347.0  1.782103 -0.588824 -0.733114  0.202365 -0.141311  0.297063   

              V7        V8        V9  ...       V21       V22       V23  \
153763  0.151686 -0.031426  0.080844  ...  0.040497  0.537483 -0.208233   
67879  -1.147468  0.212974 -0.274417  ... -0.013013  0.288307  0.009987   
190575 -0.697817  0.406668 -0.346473  ...  0.239632  0.546754  0.182530   
182469 -0.059541  0.869103  0.540991  ... -0.200652 -0.522108 -0.039645   
216102 -0.364489  0.102535  0.820113  ... -0.129255 -0.482629  0.276276   

             V24       V25       V26       V27       V28  Amount  Class  
153763  0.575130 -0.802597 -0.144597  0.207653  0.194421    7.56      0  
67879   0.362913  0.084878 -0.250990  0.097745  0.062726  100.00      0  
190575  0.707514 -0.315901 -0.028937 -0.010816 -0.026827   60.45      0  
182469  0.670408 -0.237620 -0.292700 -0.252867 -0.213708    1.46      0  
216102  0.267914 -0.419479 -0.701441  0.011196 -0.019785  107.36      0  

[5 rows x 31 columns]

In [102]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [104]:
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [106]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      98993.447154 -0.089352 -0.098597 -0.184446  0.194463 -0.017027   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.011678  0.064615 -0.042989  0.015218  ... -0.125351  0.002639   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.009967 -0.073144  0.003848 -0.032651 -0.015232 -0.013313 -0.010248   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       85.475915  
1      122.211321  

[2 rows x 30 columns]

The sampled transactions show clear differences in feature values between fraud and legitimate transactions. This helps the model learn fraud patterns better.

Splitting the data into Features & Targets

In [110]:
X = new_dataset.drop(columns='Class', axis=1)
Y = new_dataset['Class']

In [112]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
153763  100005.0 -0.947856  0.488175  0.776310  3.244927  2.302845 -0.073014   
67879    52734.0  1.030633 -0.887742  1.446677  0.582385 -1.736129  0.030491   
190575  128890.0  1.722578 -0.055197  0.100595  3.725552 -0.225874  0.941944   
182469  125411.0 -0.196458  0.239535  0.155793 -2.081468  2.489614  3.927646   
216102  140347.0  1.782103 -0.588824 -0.733114  0.202365 -0.141311  0.297063   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

X now contains only the features (V1, V2, ..., Amount, etc.).

Y contains 0 for legit transactions and 1 for fraud transactions.

In [116]:
print(Y)

153763    0
67879     0
190575    0
182469    0
216102    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Split the data into Training data & Testing Data

In [119]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify=Y, random_state=2)

Training on 80% of data helps the model learn patterns.
Testing on 20% of data helps check if the model works on unseen data.

Training Phase : 
The model learns patterns from X_train (input) and Y_train (output).


Testing Phase : 
The model makes predictions on X_test.
Predictions are compared with Y_test to check accuracy.

stratify=Y :- same proportion
random_state=2 :- split remains the same every time the code runs 

In [125]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


Model Training

Logistic Regression

 Logistic Regression is simple, efficient, and well-suited for binary classification (fraud or not fraud). It requires less computational power and performs well when combined with data balancing techniques.

In [129]:
model = LogisticRegression()

We use StandardScaler to make sure all features in the dataset have the same scale.

If we don’t scale them, bigger numbers might dominate the model.

In [159]:

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [161]:
# training the Logistic Regression Model with Training Data
model.fit(X_train, Y_train)

LogisticRegression()

X_train → Contains the input features (scaled values of transactions).
Y_train → Contains the labels (0 for legit transactions, 1 for fraud).
model.fit() → The model learns patterns from X_train to predict Y_train

Model Evaluation

Accuracy Score

Calculates accuracy as:

Accuracy= 

Correct Predictions/Total Predictions
​
 
Gives an idea of how well the model has learned from the training data.

In [166]:
# accuracy on training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [168]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9491740787801779


In [170]:
# accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [172]:
print('Accuracy score on Test Data : ', test_data_accuracy)

Accuracy score on Test Data :  0.9137055837563451


conclusion:

The Logistic Regression model achieved 94.28% training accuracy and 91.37% test accuracy, showing good performance.

The small accuracy drop indicates low overfitting